src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp

```diff
@@ -0,0 +1,1001 @@
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  ShenandoahCollectionSet* collection_set = heap->collection_set();
+  bool is_generational = heap->mode()->is_generational();
+
+  assert(!heap->is_full_gc_in_progress(), "Only for concurrent and degenerated GC");
+  assert(!is_old(), "Only YOUNG and GLOBAL GC perform evacuations");
+  {
+    ShenandoahGCPhase phase(concurrent ? ShenandoahPhaseTimings::final_update_region_states :
+                            ShenandoahPhaseTimings::degen_gc_final_update_region_states);
+    ShenandoahFinalMarkUpdateRegionStateClosure cl(complete_marking_context());
+    parallel_heap_region_iterate(&cl);
+
+    if (is_young()) {
+      // We always need to update the watermark for old regions. If there
+      // are mixed collections pending, we also need to synchronize the
+      // pinned status for old regions. Since we are already visiting every
+      // old region here, go ahead and sync the pin status too.
+      ShenandoahFinalMarkUpdateRegionStateClosure old_cl(nullptr);
+      heap->old_generation()->parallel_heap_region_iterate(&old_cl);
+    }
+  }
+
+  // Tally the census counts and compute the adaptive tenuring threshold
+  if (is_generational && ShenandoahGenerationalAdaptiveTenuring && !ShenandoahGenerationalCensusAtEvac) {
+    // Objects above TAMS weren't included in the age census. Since they were all
+    // allocated in this cycle they belong in the age 0 cohort. We walk over all
+    // young regions and sum the volume of objects between TAMS and top.
+    ShenandoahUpdateCensusZeroCohortClosure age0_cl(complete_marking_context());
+    heap->young_generation()->heap_region_iterate(&age0_cl);
+    size_t age0_pop = age0_cl.get_age0_population();
+
+    // Update the global census, including the missed age 0 cohort above,
+    // along with the census done during marking, and compute the tenuring threshold.
+    ShenandoahAgeCensus* census = ShenandoahGenerationalHeap::heap()->age_census();
+    census->update_census(age0_pop);
+#ifndef PRODUCT
+    size_t total_pop = age0_cl.get_total_population();
+    size_t total_census = census->get_total();
+    // Usually total_pop > total_census, but not by too much.
+    // We use integer division so anything up to just less than 2 is considered
+    // reasonable, and the "+1" is to avoid divide-by-zero.
+    assert((total_pop+1)/(total_census+1) ==  1, "Extreme divergence: "
+           SIZE_FORMAT "/" SIZE_FORMAT, total_pop, total_census);
+#endif
+  }
+
+  {
+    ShenandoahGCPhase phase(concurrent ? ShenandoahPhaseTimings::choose_cset :
+                            ShenandoahPhaseTimings::degen_gc_choose_cset);
+
+    collection_set->clear();
+    ShenandoahHeapLocker locker(heap->lock());
+    if (is_generational) {
+      // Seed the collection set with resource area-allocated
+      // preselected regions, which are removed when we exit this scope.
+      ResourceMark rm;
```

Comment:
You can fold `ResourceMark` into `ShenandoahCollectionSetPreselector` class.

Reply:
This makes sense: https://bugs.openjdk.org/browse/JDK-8343227